### PR TITLE
feature(core): support fill pattern image for features to raster.

### PR DESCRIPTION
**Pull request by gchoqueux**
## Description
Support fill pattern image for features (geojson) to raster.

## Screenshots

cross

### CODE REVIEW

**Comment by cboucheIGN** [comment]
This look great !

File: examples/source_file_geojson_raster.html [hunk 1]
```diff
@@ -52,39 +52,70 @@
...
-                        name: 'ariege',
-                        transparent: true,
-                        style: {
-                            fill: {
-                              color: 'orange',
-                              opacity: 0.5,
-                            },
-                            stroke: {
-                                color:'white',
+                    return itowns.Fetcher.texture('images/cross.png').then(texture => {
```
Comment (zarov) [inline examples/source_file_geojson_raster.html]: Where is `texture` used here ?

File: src/Core/Style.js [hunk 2]
```diff
@@ -26,6 +26,7 @@ class Style {
      * @param {object} [params]
      * @param {object} [params.fill] fill style.
      * @param {string} [params.fill.color] fill color string css.
+     * @param {Image|Canvas} [params.fill.pattern] fill with pattern image.
```
Comment (zarov) [inline src/Core/Style.js]: I would prefer allowing a `THREE.Texture`, then you can test it using `isTexture`.

File: src/Converter/Feature2Texture.js [hunk 3]
```diff
@@ -40,15 +42,19 @@ function _drawPolygon(ctx, vertices, indices, style, size, extent, invCtxScale,
...
         ctx.fill();
     }
 }
 
-function fillStyle(style, ctx) {
-    if (ctx.fillStyle !== style.fill.color) {
+function fillStyle(style, ctx, invCtxScale) {
+    if (style.fill.color && ctx.fillStyle !== style.fill.color) {
         ctx.fillStyle = style.fill.color;
+    } else if (style.fill.pattern && ctx.fillStyle.src !== style.fill.pattern.src) {
```
Comment (zarov) [inline src/Converter/Feature2Texture.js]: Shouldn't this be the first option to check ?

File: examples/source_file_geojson_raster.html [hunk 4]
```diff
@@ -52,39 +52,70 @@
...
+                        var ariegeLayer = new itowns.ColorLayer('pyrenees-orientales', {
+                            name: 'pyrenees-orientales',
+                            transparent: true,
+                            style: {
+                                fill: {
+                                  pattern: texture.image,
+                                  opacity: 0.8,
+                                },
+                                stroke: {
+                                    color:'IndianRed',
```
Comment (zarov) [inline examples/source_file_geojson_raster.html:109]: How is this affecting the visual ? Is it a fallback in case the image isn't available ?

**Review by zarov — COMMENTED** [review]
Really nice feature to have, way less complicated than I thought it could be ! You should consider expanding a little more the `Style` documentation with examples, or even consider writing a tutorial about customisation.

File: examples/source_file_geojson_raster.html (hunk 1 above)
Comment (gchoqueux) [reply examples/source_file_geojson_raster.html]: I use `texture fetcher` to load image, by using `texture.image`.

Comment (zarov) [reply examples/source_file_geojson_raster.html]: Yes, but you don't use it for this layer, so you can remove it.

File: src/Converter/Feature2Texture.js (hunk 3 above)
Comment (gchoqueux) [reply src/Converter/Feature2Texture.js]: yes

File: src/Core/Style.js (hunk 2 above)
Comment (gchoqueux) [reply src/Core/Style.js]: Canvas context couldn't use `THREE.Texture`.

File: examples/source_file_geojson_raster.html (hunk 4 above)
Comment (gchoqueux) [reply examples/source_file_geojson_raster.html:109]: No, it's to color the stroke.

File: src/Core/Style.js (hunk 2 above)
Comment (zarov) [reply src/Core/Style.js]: Yes, but below you can then use `params.fill.pattern.image`. This way the texture could be updated without having to change the style ?

File: examples/source_file_geojson_raster.html (hunk 4 above)
Comment (zarov) [reply examples/source_file_geojson_raster.html:109]: I can see that in the screenshot yes, sorry.

File: examples/source_file_geojson_raster.html (hunk 1 above)
Comment (gchoqueux) [reply examples/source_file_geojson_raster.html]: ok

**Review by zarov — APPROVED** [review]
LGTM, thanks

**Pull request merged**